### PR TITLE
Fake source integration testing

### PR DESCRIPTION
**Pull request by jzohrab**
## Summary

This adds the idea of "fake sources" for integration testing of Li itself.

We need a way to ensure that some base functions will always work, without actually being tied to real sources, which can go up or down.  The fake sources can scrape localhost, provided by the sandbox, so we can guarantee that these will always work.  This will also let us double-check that the code can handle cases which might be hard to test for in live sources.

## Changes

* adds a `LI_SOURCES_DIR` env var to use fake sources for integration tests, which sourceMap respects
* adds a fake scraper and a simple test to verify crawl and scrape.
* cleans up sandbox startup and shut down during tests

## Verifying

`TEST_ONLY=kr npm run test:integration` - passes
`npm run test` - passes
`./start --crawl kr` - works
`./start --scrape kr` - works

### CODE REVIEW

File: tests/integration/_lib/fake-crawl-sites.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+const fs = require('fs')
+
+/** Folder under sandbox public that the sources scrape.
+ *
+ * These can be crawled with URLs like
+ * http://localhost:5555/tests/fake-source-urls/fake/fake.json. */
+const baseFolder = join(process.cwd(), 'public', 'tests', 'fake-source-urls')
+
+/** Write a test file to be scraped. */
+function writeFile (subdir, filename, content) {
```
Comment (ryanblock) [inline tests/integration/_lib/fake-crawl-sites.js:11]: Why not just `module.exports = function writeFile`?

File: public/.gitignore [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Localhost acts as a crawl source for fake sources during integration
+# tests.  See tests/integration/fake-sources.
+
+/tests
```
Comment (ryanblock) [inline public/.gitignore:4]: We should not put files in public/ that are not of our production deployment artifacts

File: tests/integration/_lib/sandbox.js [hunk 3]
```diff
@@ -34,10 +53,17 @@ async function stop () {
...
+  console.log('All tests done.  Shutting down sandbox ...')
+  _stop()
+  console.log('Sandbox shut down.')
+
+})
+
 module.exports = {
-  start,
-  stop
+  start
```
Comment (ryanblock) [inline tests/integration/_lib/sandbox.js:68]: Same as above, why exporting in an object?

File: tests/integration/fake-sources/fake.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+module.exports = {
+  country: 'iso1:US',
+  state: 'CA',
+  maintainers: [],
+  friendly: {
+    name: 'Fake source',
+    url: 'rick-roll-url'
```
Comment (ryanblock) [inline tests/integration/fake-sources/fake.js:7]: I mean, come on, put the actual URL in there at least! 🎤

File: tests/integration/_lib/fake-crawl-sites.js (hunk 1 above)
Comment (jzohrab) [reply tests/integration/_lib/fake-crawl-sites.js:11]: I have a couple of ideas in mind for future things -- eg small test fixtures -- so we might want to do something like `function copyFixture(subdir, fixturePath)` later.

File: public/.gitignore (hunk 2 above)
Comment (jzohrab) [reply public/.gitignore:4]: Got it, I'll move this rule up to the root gitignore.  The tests copy things to public, and then remove them.

File: tests/integration/_lib/sandbox.js (hunk 3 above)
Comment (jzohrab) [reply tests/integration/_lib/sandbox.js:68]: Originally there were other functions, but now that it's only start I'll switch this.  Thanks!

File: public/.gitignore (hunk 2 above)
Comment (jzohrab) [reply public/.gitignore:4]: Ignore rule moved up to `./.gitignore`, thanks!

File: tests/integration/fake-sources/fake.js (hunk 4 above)
Comment (jzohrab) [reply tests/integration/fake-sources/fake.js:7]: Done!

File: tests/integration/_lib/sandbox.js (hunk 3 above)
Comment (jzohrab) [reply tests/integration/_lib/sandbox.js:68]: I tried changing this to module.exports in various ways, but I'm calling `start()` right after this definition, and with various permutations of this (2 sample tries below), things either failed or felt clumsy.  I'll leave it as-is, but if there's a better way, please let me know!

## Try 1 -- `module.exports = async function start () { ... }`

doesn't work

| tried | error |
| --- | --- |
| `start()` | `ReferenceError: start is not defined` |
| `exports.start()` | `exports.start is not a function` |

## Try 2 -- `module.exports.start = async function () {...}`

This works, but looks clumsy to me:

```
module.exports.start = async function () {...}
exports.start()
```

Comment (jzohrab) [reply tests/integration/_lib/sandbox.js:68]: Open to suggestions on this one, perhaps there's a sharp way to do it.

File: public/.gitignore (hunk 2 above)
Comment (jzohrab) [reply public/.gitignore]: Also added delete on sandbox stop, JIC.

File: tests/integration/_lib/sandbox.js (hunk 3 above)
Comment (ryanblock) [reply tests/integration/_lib/sandbox.js:79]: I'll take a look

File: src/shared/sources/_lib/source-map.js [hunk 5]
```diff
@@ -3,8 +3,9 @@ const globJoin = require('../../utils/glob-join.js')
 const sourceKey = require('./source-key.js')
 const { sep } = require('path')
 
-module.exports = function sourceMap () {
-  const scrapers = globJoin(__dirname, '..', '**', '*.js')
+module.exports = function sourceMap (params = {}) {
+  const scrapersPath = params._sourcesPath || globJoin(__dirname, '..')
+  const scrapers = globJoin(scrapersPath, '**', '*.js')
   let filePaths = glob(scrapers)
```
Comment (ryanblock) [inline src/shared/sources/_lib/source-map.js:9]: idk why I wrote this originally as scrapers but shouldn't this probably be named sources? (Law of triviality coming into effect here...)

File: tests/integration/events/scraper-test.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  t.equal('iso1:us#iso2:us-ca', result.locationIDs.join(), 'Location IDs')
+
+  const actual = result.data
+  t.ok(actual, 'Have data')
+  t.equal(1, actual.length, '1 record in returned data')
+
+  // These fields should match exactly.
+  const expected = {
+    cases: 10,
+    deaths: 20,
```
Comment (ryanblock) [inline tests/integration/events/scraper-test.js:45]: Should probably add more fields, max out the expected output so we can make sure *nothing* gets lost along the way.

File: src/shared/sources/_lib/source-map.js (hunk 5 above)
Comment (jzohrab) [reply src/shared/sources/_lib/source-map.js:9]: grrrr yes it should, thanks!

File: tests/integration/events/scraper-test.js (hunk 6 above)
Comment (jzohrab) [reply tests/integration/events/scraper-test.js:45]: I'm not sure what fields we can include, so this is a first iteration. :-)

File: src/shared/sources/_lib/source-map.js (hunk 5 above)
Comment (jzohrab) [reply src/shared/sources/_lib/source-map.js]: Fixed.

File: tests/integration/events/scraper-test.js (hunk 6 above)
Comment (jzohrab) [reply tests/integration/events/scraper-test.js:45]: Let's add a TODO for this one.

Comment (jzohrab) [reply tests/integration/events/scraper-test.js:46]: TODO added.

**Pull request merged**